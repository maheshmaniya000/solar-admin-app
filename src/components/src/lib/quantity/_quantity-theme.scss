@use '~@angular/material' as mat;

@mixin theme($theme-or-color-config) {
  $typography-config: mat.get-typography-config($theme-or-color-config);
  $color-config: mat.get-color-config($theme-or-color-config);
  $foreground: map-get($color-config, foreground);

  sg-quantity {
    .title,
    .sub-title,
    .unit {
      @include mat.typography-level($typography-config, caption);
    }

    .title,
    .value {
      color: mat.get-color-from-palette($foreground, text);
    }

    .sub-title,
    .unit {
      color: mat.get-color-from-palette($foreground, hint-text);
    }

    .value {
      @include mat.typography-level($typography-config, display-1);
      font-weight: mat.font-weight($typography-config, display-2);
    }
  }
}
