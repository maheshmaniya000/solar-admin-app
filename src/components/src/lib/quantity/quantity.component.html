<ng-container
  *ngTemplateOutlet="
    titleTemplate;
    context: { $implicit: { className: 'title', title: title } }
  "
></ng-container>

<ng-container
  *ngTemplateOutlet="
    titleTemplate;
    context: { $implicit: { className: 'sub-title', title: subTitle } }
  "
></ng-container>

<div>
  <span class="value">{{ value }}</span>
  <span class="unit" [innerHTML]="unitHtml"></span>
</div>

<ng-template #titleTemplate let-config>
  <div
    *ngIf="config.title"
    [class]="config.className"
    [style]="{ '-webkit-line-clamp': config.title.lineCount }"
    [style.height.px]="calculateTitleHeight(config.title.lineCount)"
  >
    {{ config.title.text }}
  </div>
</ng-template>
