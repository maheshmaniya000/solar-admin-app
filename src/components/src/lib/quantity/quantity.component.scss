:host {
  display: inline-block;
}

.title,
.sub-title {
  display: -webkit-box;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
}

.unit {
  padding-left: 8px;
}
