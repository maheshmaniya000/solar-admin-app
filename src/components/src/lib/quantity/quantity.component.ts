import { ChangeDetectionStrategy, Component, Input } from '@angular/core';

import { SgScssVariables } from '../../styles/scss-variables';
import { isNaturalNumberOrZero } from '../utils';

interface QuantityTitle {
  text: string;
  lineCount: number;
}

@Component({
  selector: 'sg-quantity',
  templateUrl: './quantity.component.html',
  styleUrls: ['./quantity.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class QuantityComponent {
  @Input() title: QuantityTitle;
  @Input() subTitle: QuantityTitle;
  @Input() value: number;
  @Input() unitHtml: string;

  calculateTitleHeight(lineCount: number): number {
    if (!isNaturalNumberOrZero(lineCount)) {
      throw new Error('lineCount must be a natural number or zero');
    }

    return SgScssVariables.sgCaptionLineHeight * lineCount;
  }
}
