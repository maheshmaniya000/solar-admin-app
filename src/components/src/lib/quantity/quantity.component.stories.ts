import { moduleMetadata, Story } from '@storybook/angular';

import { QuantityModule } from './quantity.module';

const createTemplate = (
  titlePropName = 'title',
  subTitlePropName = 'subTitle',
  valuePropName = 'value',
  unitHtmlPropName = 'unitHtml'
): string => `
  <sg-quantity
    [title]="${titlePropName}"
    [subTitle]="${subTitlePropName}"
    [value]="${valuePropName}"
    [unitHtml]="${unitHtmlPropName}">
  </sg-quantity>
`;

const styles = [
  `
    sg-quantity {
      width: max(150px, 15%);
      padding: 10px;
    }
  `
];

export default {
  title: 'Solargis/Components/Quantity',
  decorators: [
    moduleMetadata({ imports: [QuantityModule] })
  ]
};

export const basic: Story = () => ({
  template: createTemplate(),
  props: {
    title: { text: 'Air temperature', lineCount: 1 },
    subTitle: { text: 'Yearly average', lineCount: 1 },
    value: 25.4,
    unitHtml: 'Â°C'
  },
  styles
});

export const multiline: Story = () => ({
  template: `
    <div class="wrapper" data-test="multiline-quantity-story">
      ${createTemplate('title1', 'subTitle1', 'value1', 'unitHtml1')}
      ${createTemplate('title2', 'subTitle2', 'value2', 'unitHtml2')}
      ${createTemplate('title3', 'subTitle3', 'value3', 'unitHtml3')}
      ${createTemplate('title4', 'subTitle4', 'value4', 'unitHtml4')}
    </div>
  `,
  props: {
    title1: { text: 'Specific photovoltaic power output yield', lineCount: 2 },
    subTitle1: { text: 'Yearly average', lineCount: 1 },
    value1: 940.1,
    unitHtml1: 'kWh/kWp',

    title2: { text: 'Global horizontal irradiation', lineCount: 1 },
    subTitle2: {
      text: 'Average monthly sum of global horizontal irradiation',
      lineCount: 2
    },
    value2: 153,
    unitHtml2: 'Wh/m<sup>2</sup>',

    title3: { text: 'Capacity factor', lineCount: 3 },
    value3: 7.4,
    unitHtml3: '%',

    title4: {
      text:
        'Total system performance considering technical availability and losses due to snow',
      lineCount: 3
    },
    value4: 1177.7,
    unitHtml4: 'kWh/kWp'
  },
  styles: styles.concat([
    `
      .wrapper {
        display: flex;
        flex-wrap: wrap;
      }
    `
  ])
});

export const interactive: Story = props => ({
  props,
  template: createTemplate(),
  styles
});

interactive.argTypes = {
  title: {
    name: 'Title',
    control: { type: 'object' },
    defaultValue: {
      text: 'Global horizontal irradiation',
      lineCount: 1
    }
  },
  subTitle: {
    name: 'Sub title',
    control: {type: 'object'},
    defaultValue: {
      text:
        'Average annual, monthly or daily sum of global horizontal irradiation',
      lineCount: 1
    }
  },
  value: {
    name: 'Value',
    control: {type: 'number'},
    defaultValue: 153
  },
  unitHtml: {
    name: 'Unit',
    control: {type: 'text'},
    defaultValue: 'Wh/m<sup>2</sup>'
  }
};
