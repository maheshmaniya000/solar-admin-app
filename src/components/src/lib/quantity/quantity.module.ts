import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { QuantityComponent } from './quantity.component';

@NgModule({
  declarations: [QuantityComponent],
  imports: [CommonModule],
  exports: [QuantityComponent]
})
export class QuantityModule {}
