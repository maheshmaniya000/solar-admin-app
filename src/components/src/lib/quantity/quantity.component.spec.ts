import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';

import { QuantityComponent } from './quantity.component';

describe('QuantityComponent', () => {
  const createComponentInputs = (): Partial<QuantityComponent> => ({
    title: { text: 'some title', lineCount: 1 },
    subTitle: { text: 'some sub title', lineCount: 2 },
    value: 129.37,
    unitHtml: 'km'
  });

  describe('unit', () => {
    let component: QuantityComponent;

    beforeEach(() => {
      component = new QuantityComponent();
    });

    describe('calculateTitleHeight', () => {
      [undefined, null, -5, 1.3].forEach(value =>
        it(`should throw error when ${value} is provided`, () =>
          expect(() => component.calculateTitleHeight(value)).toThrow())
      );

      it('should return the title height based on the provided number of lines it requires', () => {
        expect(component.calculateTitleHeight(0)).toEqual(0);
        expect(component.calculateTitleHeight(2)).toEqual(36);
      });
    });
  });

  describe('component', () => {
    let spectator: Spectator<QuantityComponent>;

    const createComponent = createComponentFactory({
      component: QuantityComponent
    });

    beforeEach(() => {
      spectator = createComponent({ props: createComponentInputs() });
    });

    it('should render with correctly bound attributes and styles', () => {
      expect(spectator.fixture).toMatchSnapshot();
    });

    it('should render simple html elements that are part of the unit property', () => {
      spectator.setInput({ unitHtml: 'Wh/m<sup>2</sup>' });
      expect(spectator.fixture).toMatchSnapshot();
    });
  });
});
