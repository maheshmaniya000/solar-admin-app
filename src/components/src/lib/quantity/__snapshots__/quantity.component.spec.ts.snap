// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuantityComponent component should render simple html elements that are part of the unit property 1`] = `
<sg-quantity
  __ngContext__={[Function LRootView]}
  subTitle={[Function Object]}
  title={[Function Object]}
  unitHtml={[Function String]}
  value={[Function Number]}
>
  <div
    class="title"
    style="height: 18px;"
  >
     some title 
  </div><div
    class="sub-title"
    style="height: 36px;"
  >
     some sub title 
  </div><div>
    <span
      class="value"
    >
      129.37
    </span>
    <span
      class="unit"
    >
      Wh/m
      <sup>
        2
      </sup>
    </span>
  </div>
</sg-quantity>
`;

exports[`QuantityComponent component should render with correctly bound attributes and styles 1`] = `
<sg-quantity
  __ngContext__={[Function LRootView]}
  subTitle={[Function Object]}
  title={[Function Object]}
  unitHtml={[Function String]}
  value={[Function Number]}
>
  <div
    class="title"
    style="height: 18px;"
  >
     some title 
  </div><div
    class="sub-title"
    style="height: 36px;"
  >
     some sub title 
  </div><div>
    <span
      class="value"
    >
      129.37
    </span>
    <span
      class="unit"
    >
      km
    </span>
  </div>
</sg-quantity>
`;
